### PR TITLE
Added time zone management module

### PR DESCRIPTION
**Pull request by PeterConijn**
This module aims to provide methods to correctly calculate the offset between a datetime in a time zone from UTC and the offset between time zones. Other than the functions in the Type Helper codeunit in the base app, this module takes daylight saving time into account when applicable.

Additionally, the module provides indicators on whether a time zone supports daylight saving time or whether a given datetime falls within the daylight saving time period for a given time zone.

### CODE REVIEW

File: Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al
Comment (pri-kise) [inline Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al:21]: The TimeZoneId is missing in this error

Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al]: Fixed missing TimeZoneId.

**Comment by pri-kise** [comment]
@PeterConijn your pull request looks very promising.

It would be really useful in my opinion to provide a procedure that uses the current time zone to calculate the offset.
Additionally, have you thought about adding a wrapper for `GetSystemTimeZones` to give access to the available time zone ids?
(https://learn.microsoft.com/en-us/previous-versions/windows/embedded/ms912391(v=winembedded.11)?redirectedfrom=MSDN)
If I read the documentation correctly these names must not be always the same.
https://learn.microsoft.com/de-de/dotnet/api/system.timezoneinfo.getsystemtimezones?view=net-7.0

Now waiting for someone of microsoft.

**Comment by PeterConijn** [comment]
> It would be really useful in my opinion to provide a procedure that uses the current time zone to calculate the offset.

That is a good idea. I will look into it and add it to the PR

> Additionally, have you thought about adding a wrapper for `GetSystemTimeZones` to give access to the available time zone ids?

A list of time zones is already handled by the Time Zone Selection module in the System App so it would be superfluous.

**Comment by pri-kise** [comment]
@PeterConijn thanks for clarification. I must have missed this System Feature.

I guess you have to include your permissionset in on the system application basic permissonset

**Comment by NKarolak** [comment]
@pri-kise, as Peter indicated: Once this PR is merged, please remember to obsolete/move TypeHelper functions to the new module. Old code:
https://github.com/StefanMaron/MSDyn365BC.Code.History/blob/bcbdcfc434259c1bfc3708b55d62752c5b5fca26/BaseApp/Source/Base%20Application/TypeHelper.Codeunit.al#L328

When you search the Base app for further occurrences of "TimeZone", you will find even more procedures to obsolete, such as https://github.com/StefanMaron/MSDyn365BC.Code.History/blob/1c682b0af15e55f729253392b821e64244712e68/BaseApp/Source/Base%20Application/TransformationRule.Table.al#L378 (provided such function exists in the new module; I did not check).

File: Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al
Comment (pri-kise) [inline Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al:19]: Signature of procedure should be updatet to only need the SourceDateTime

Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al:19]: Shows you should never code while on the phone. 

I fixed the signature and added some other minor fixes to the PR.

**Comment by PeterConijn** [comment]
The automated builds seem to fail on the object nos (I used the default 50100-50149 range). Maybe a noobie question, but is this something Microsoft will do or is that something expected of me? If the latter, I would like to know which object numbers I can use.

**Comment by pri-kise** [comment]
The best way is to wait for someone of microsoft and update the id's for the Tests and the new Objects afterwards.

**Comment by pri-kise** [comment]
@JesperSchulz could you please assign someone to this pull request?
To give feedback if this module can be accepted and to provide some new Ids, that can be used for this new module?

**Comment by JesperSchulz** [comment]
I was hoping for other members of the community to review this code too ;-) I know this particular repository has a history of "partner develops, Microsoft reviews", but in the long run that won't scale. This PR is something that many partners over time have requested, and I was hoping to see some involvement from them.
Oh yeah, and it's year-end, which always means "go-time" for all Business Central developers - that may also play into, why we haven't looked at this PR yet ourselves ;-) But I promise to do so today/tomorrow! The ID problem is an easy one to fix - that I can do for sure. Stay tuned!

**Comment by pri-kise** [comment]
@JesperSchulz do you have some Ids for the Test Codeunits too?

**Comment by JesperSchulz** [comment]
132979 seems to be available, both for the PS and test :-)

**Comment by PeterConijn** [comment]
> @PeterConijn, if you grant me permissions to your repo, I can push the new IDs :-) Alternatively, you can update them yourself. I've reserved 8720 - 8724 for you.

Thanks @JesperSchulz. I have done both. You have contributor access to my fork and I have updated the object numbers in the app, test, and test permissions.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (pri-kise) [inline Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:1]: This PermissionSet should probably use the Id of the normal app `8720`.
The ID Range: 
```json
{
      "from": 132979,
      "to": 132979
    }
```
is only meant for testapps

**Comment by JesperSchulz** [comment]
Great! Now let's see if this builds, and then it's time for code review 🥳

**Comment by JesperSchulz** [comment]
Build failed. Here's the error: AL0185 Codeunit 'Library - Test Initialize' is missing

Please make build pass before we continue code review.

**Comment by PeterConijn** [comment]
Thanks for the reminder! 

Fixed the bug. Build succeeded.

**Review by adrogin — APPROVED** [review]

**Review by pri-kise — DISMISSED** [review]

File: Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al
Comment (miljance) [inline Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al:21]: Can the Session have empty in TimeZone? If yes what happens in this case - Duration will be 0? If no is this line really needed?

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (miljance) [inline Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:1]: I think that the permission set should be included in objects PermissionSet 219 "System Application - Objects". Otherwise this permission set is not accessible nor assignable.

**Review by miljance — CHANGES_REQUESTED** [review]
Could you please include the permission set as noted in the feedback?

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:5]: Let's add a blank line here between assignable and permissions - just to make the permissions portion stand out a little more.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:1]: This permission set should probably be called "Time Zone Management - Objects", just like the caption. This one should be internal. You could then create a new permission set, which isn't internal and has the name "Time Zone Management - Execute". For now, that permission set would only contain the objects one...

File: Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al:1]: As this is a public codeunit you must add:
```
/// <summary>
/// write me
/// <example>
write me
/// </example>
/// </summary>
```

Also, you need to add the open source header to all objects:
```
// ------------------------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
// ------------------------------------------------------------------------------------------------
```

File: Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al:9]: Looks like we need a dependency to the dotnet module.

File: Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al:1]: We usually don't CamelCase object names.

File: Modules/System/Time Zone Management/app.json
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/app.json:3]: I wonder if we really need a module for "Time Zone Management", or if it maybe would make sense to make the module a tad wider? We've got a lot of DateTime conversion functions in the BaseApp - maybe these could find a new home in this module too?

File: Modules/System/Time Zone Management/app.json
Comment (JesperSchulz) [inline Modules/System/Time Zone Management/app.json:2]: We've tried to stop calling things "management", as it really doesn't add much information to module. "Time Zone" would be fine, I guess?

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (JesperSchulz) [inline Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al:205]: Let's get all of the global variables moved to the top. It's kind of a standard we have.

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (JesperSchulz) [inline Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al:266]: These local helper methods seem only the called once in the tests - why have them in the first place?

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (JesperSchulz) [inline Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al:313]: If we've being honsest with ourselves, the name of this function should not be GetTimeZoneIdForNonDSTTimeZone(). It should rather be GetTimeZoneIdForUSMountainStandardTime. You are using a lot of hard coded values and "magic numbers" in your test - that could be improved.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (pri-kise) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:1]: @JesperSchulz  Keep in mind, that PermissionSets should have a max length of 20 chars.

File: Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al:21]: I believe it can be empty when called from a web service, for example. See procedures GetTimezoneOffset and GetUserClientTypeOffset for similar patterns.

That said, if I remove this statement, the missing time zone will be handled with an error in the initializer codeunit. That is a choice that we will have to make. Given the code in codeunit 10, I opted for similar patterning as in that codeunit.

File: Modules/System/Time Zone Management/app.json
Comment (pri-kise) [reply Modules/System/Time Zone Management/app.json:3]: @JesperSchulz What are thinking of?
Something like "Date and Time" or "Date & Time Utility", ...
And if this becomes whole Date,DateTime,DateTimeOffset,TimeZone Module. Are the Id's still the correct ones or should they be changed to some where we can easily add several more objects without the need for additional Id Ranges?

File: Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/DateTimeOffset.Codeunit.al:9]: Added the MIT license header and the summary.

File: Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/TimeZoneInfoInitializer.Codeunit.al]: I did that in this case, since the DotNet object being initialized is TimeZoneInfo, but I have added spaces.

File: Modules/System/Time Zone Management/app.json
Comment (PeterConijn) [reply Modules/System/Time Zone Management/app.json:2]: Perhaps, but given your previous comment where you would like to add codeunit 10 BaseApp functionality to it, would it be better to resolve this once agreement has been reached there?

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (PeterConijn) [reply Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al:274]: It keeps the test procedures better structured in the given-when-then pattern, thus increasing readability.

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (PeterConijn) [reply Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al:321]: True, but the problem with this sort of functionality is that it does not exist yet. In a good test, you would have two ways of reaching the same result: the one you want to test and the control where you know the answer is correct.

The problem here is that there is no real alternative method to reach the correct offset; if there were, I would not have needed to create this module. The only thing I can do is use the same method using the TimeZoneMgt dotnet module, which would reliably use the same  time.

A possible alternative is using several time zones in the set, but I don't really see a way around using hard-coded information.

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (PeterConijn) [reply Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al]: Yes, you are right, of course. My test codeunit template puts them in that spot and I forgot to move them. Did that now.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (JesperSchulz) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:6]: 30, I believe? We've got plenty with a length of more than 20 chars (e.g. permissionset 9010 "AAD User Management - Objects").

File: Modules/System/Time Zone Management/app.json
Comment (JesperSchulz) [reply Modules/System/Time Zone Management/app.json:3]: Exactly, a "Date & Time" module. That'd be lovely. And no, we would not have to change the ID's. While ranges are preferable, they're not a requirement. We can pick individual IDs.

File: Modules/System/Time Zone Management/app.json
Comment (JesperSchulz) [reply Modules/System/Time Zone Management/app.json:2]: Absolutely.

Comment (JesperSchulz) [reply Modules/System/Time Zone Management/app.json:2]: So, what should we agree upon? Should we make a "Date & Time" module, which contains a Timezone facade and implementation codeunit, containing both DateTimeOffset and DaylightSavingTimeInfo? It would be a more versatile module.
Input to that idea?

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (pri-kise) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:6]: I recieve the following warning in VS Code.
image
Maybe this warning is outdated.
But there has been some limitation.
At least the Role ID in`table 2000000167 "Aggregate Permission Set"` is limited to Code[20]
and the Name is limited to Text[30].
Therefore we always add a maxlength = 30 to the PermissionSet Caption

```al
table 2000000167 "Aggregate Permission Set"
{
    DataPerCompany = false;
    Scope = Cloud;

    //WriteProtected=True;
    fields
    {
        field(1; Scope; Option)
        {
            OptionMembers = System,Tenant;
        }
        field(2; "App ID"; Guid)
        {
        }
        field(3; "Role ID"; Code[20])
        {
        }
        field(4; Name; Text[30])
        {
        }
        field(5; "App Name"; Text[250])
        {
        }
    }
    keys
    {
        key(pk; Scope, "App ID", "Role ID")
        {
        }
    }

    fieldgroups
    {
        fieldgroup("DropDown"; "Role ID", Name, "App Name")
        {

        }
    }
}
```

File: Modules/System/Time Zone Management/app.json
Comment (PeterConijn) [reply Modules/System/Time Zone Management/app.json:2]: I like the Date & Time module idea. 

To summarize and make sure we're on the same page: it will have a single timezone facade and implementation codeunit that merges the functionality that is now contained in the two separate facade/implementation pairs.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al:6]: That would make sense in and of itself. @JesperSchulz, you indicated below that this should have its own public permission set.

> This permission set should probably be called "Time Zone Management - Objects", just like the caption. This one should be internal. You could then create a new permission set, which isn't internal and has the name "Time Zone Management - Execute". For now, that permission set would only contain the objects one...

Waiting for feedback on this one.

File: Modules/System/Time Zone Management/app.json
Comment (PeterConijn) [reply Modules/System/Time Zone Management/app.json:3]: Since we are having two similar conversations on this topic, I am marking this one as resolved and we can add any details on the one below this.

File: Modules/System/Time Zone Management/app.json
Comment (JesperSchulz) [reply Modules/System/Time Zone Management/app.json:2]: Correct. We're exactly on the same page! I like that idea too. It has even grown on me since yesterday :-) Let's do that!

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al]: My compiler keeps crashing on the app.json (accessviolation error) so my code analyzer breaks like mad, but I have updated the name.

File: Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al]: Added it.

File: Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/src/DateTimeOffsetImpl.Codeunit.al]: I have added the permission set to "System Application - Objects"; this seems the most logical place.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al]: After thinking this over, I agree with @miljance that adding it to the "System Application - Objects" would suffice. A separate assignable time zone permission seems impractical.

File: Modules/System/Time Zone Management/app.json
Comment (PeterConijn) [reply Modules/System/Time Zone Management/app.json:2]: Renamed the module and merged the face and implementation as discussed.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (JesperSchulz) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al]: About permission sets... from 2023 wave 1, we're going to implement inherent permissions on these kinds of modules (permissions defined in the objects, which aren't controllable). Nobody wants to control permissions on "Time Zone" calculations - everyone should just be able to use that code. Hence, please do put it into "System Application - Objects", and I will convert the permissions to inherent permissions in the objects, when porting the code to 2023 Wave 1.

Comment (PeterConijn) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al]: Makes sense. I have added it to "System Application - Objects".

File: Modules/System/Permissions/SystemApplicationObjects.PermissionSet.al
Comment (PeterConijn) [reply Modules/System/Permissions/SystemApplicationObjects.PermissionSet.al]: The weird thing is that I remember staring at that list of permissions for that specific reason and my brain somehow forgot how the alphabet worked.

Fixed!

Comment (SHermann73) [reply Modules/System/Permissions/SystemApplicationObjects.PermissionSet.al]: That was fast! :-)
Great, thanks.

File: Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al
Comment (PeterConijn) [reply Modules/System/Time Zone Management/Permissions/TimeZoneMgtObjects.PermissionSet.al]: Permissions added to System Application - Objects

File: Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al
Comment (PeterConijn) [reply Modules/System Tests/Time Zone Management/src/TimeZoneManagementTest.Codeunit.al]: I have made some changes to the test feature to introduce a wider array of time zones to test and a certain degree of randomness in order to make the test more robust.

**Comment by JesperSchulz** [comment]
Alright, that should do it. I made a few changes to the module locally (e.g. removed permissions and used inherent permissions instead). This module will ship with v. 22 in April!

**Comment by PeterConijn** [comment]
> Alright, that should do it. I made a few changes to the module locally (e.g. removed permissions and used inherent permissions instead). This module will ship with v. 22 in April!

That's great news! Thank you very much for your collaboration in this module.

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 52216.

**Comment by JesperSchulz** [comment]
Merging the PR, even though it means the changes will get reverted with 21.4, as the module doesn't release until 22.0. But don't worry! The module will be back 🥳

**Review by JesperSchulz — APPROVED** [review]
Approved. I've made some further changes in the internal PR, but most of the code stays as it is in this PR :-)

**Pull request merged**
